node/node.go

```diff
@@ -1295,22 +1294,34 @@ type validatedBlock struct {
 	vb *ledgercore.ValidatedBlock
 }
 
-// WithSeed satisfies the agreement.ValidatedBlock interface.
-func (vb validatedBlock) WithSeed(s committee.Seed) agreement.ValidatedBlock {
-	lvb := vb.vb.WithSeed(s)
-	return validatedBlock{vb: &lvb}
+// unfinishedBlock satisfies agreement.UnfinishedBlock
+type unfinishedBlock struct {
```

Comment:
nit: instead of an extra copy of unfinishedBlock it is possible to use func types and anonymous struct in `AssembleBlock` return like this:

```go
type roundFunc func() basics.Round
func (rf roundFunc) Round() basics.Round { return rf() }
type finishBlockFunc func(s committee.Seed, proposer basics.Address, eligible bool) bookkeeping.Block
func (ff finishBlockFunc) FinishBlock(s committee.Seed, proposer basics.Address, eligible bool) agreement.ProposableBlock {
	return proposableBlock{blk: ff(s, proposer, eligible)}
}
```

AssembleBlock's return statement becomes
```go
	return struct {
		roundFunc
		finishBlockFunc
	}{
		roundFunc: func() basics.Round { return ub.Round() },
		finishBlockFunc: func(s committee.Seed, proposer basics.Address, eligible bool) bookkeeping.Block {
			return ub.FinishBlock(s, proposer, eligible)
		},
	}, nil
```

a bit hairy but no one more copy of unfinishedBlock

Reply:
Since the block's payset is not copied, I'm not sure how much we care about avoiding the copy. It's considerably smaller tha the consensus parameters we copy all over the place.

On the other hand, I think this is exactly one of the types that I don't think should exist. I think `node.unfinishedBlock` could simply not exist, and have `ledgercore.UnfinishedBlock` implement `agreement.UnfinishedBlock`.

Reply:
See if you like https://github.com/jannotti/go-algorand/commit/a46890d3a4f3d67bc2d0c13b69972051748d58cf

FinishBlock essentially becomes a cast which is a "copy", but only like any function call would be.

Reply:
It looks like the proposal makes possible to use Block everywhere and vanishes the proposable/unfinished block compile-time misuse.

Reply:
I'm going to take that as at least mildly positive (unless I'm reading your comment wrong), I'll merge it in.

Reply:
oh, sorry I meant another way around, I like the current stronger typing and not keen of weaker abstraction.

Reply:
Ok, I was afraid of that. Let me try to convince you. I think the proposed change is _more_ stronger typed than the current approach, not less.

In the proposed change, `agreement.Block` is a distinct concrete type, it is not a `bookkeeping.Block`.  You can only make an `agreement.Block` by explicit conversion. Those conversions tend to appear in `FinishBlock()` implementations.

In contrast, the existing `agreement.ProposableBlock` is an interface that only contains a `Block()` method. So any type that is an `agreement.ValidatedBlock` such as `ledgercore.ValidatedBlock` is also a ProposableBlock. It can be passed as an `agreement.ProposableBlock` without explicit conversion.

Reply:
So `agreement.Block` is a replacement for `agreement.ProposableBlock`
What confused me is  this cast `e := bookkeeping.Block(blk)` in `makeProposalFromProposableBlock` but I do agree, this is not worse than the `e := blk.Block()` from before.
Thanks for pointing out, lets apply the change.